src/fu-agent.c

```diff
@@ -0,0 +1,645 @@
+	return g_build_filename (path, "agent.conf", NULL);
+}
+
+static GKeyFile *
+fu_util_agent_get_config (GError **error)
+{
+	g_autofree gchar *fn = fu_util_agent_get_config_fn ();
+	g_autoptr(GKeyFile) config = g_key_file_new ();
+	if (!g_key_file_load_from_file (config, fn, G_KEY_FILE_KEEP_COMMENTS, error)) {
+		g_prefix_error (error, "failed to load %s: ", fn);
+		return FALSE;
+	}
+	return g_steal_pointer (&config);
+}
+
+static gboolean
+fu_util_agent_set_server (FuUtilPrivate *priv, const gchar *server, GError **error)
+{
+	g_autofree gchar *fn = fu_util_agent_get_config_fn ();
+	g_autoptr(GKeyFile) config = fu_util_agent_get_config (error);
+	if (config == NULL)
+		return FALSE;
+	g_key_file_set_string (config, "fwupdagent", "Server", server);
+	return g_key_file_save_to_file (config, fn, error);
+}
+
+static gboolean
+fu_util_agent_run_action (FuUtilPrivate *priv,
+			  JsonObject *json_object,
+			  GError **error)
+{
+	const gchar *checksum;
+	const gchar *device_id;
+	const gchar *task;
+	g_autoptr(FwupdDevice) device = NULL;
+	g_autoptr(FwupdRelease) rel = NULL;
+	g_autoptr(GPtrArray) releases = NULL;
+
+	/* check object for args */
+	if (!json_object_has_member (json_object, "Task")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "No task specified");
+		return FALSE;
+	}
+	if (!json_object_has_member (json_object, "DeviceId")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "No DeviceId specified");
+		return FALSE;
+	}
+	if (!json_object_has_member (json_object, "Checksum")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "No checksum specified");
+		return FALSE;
+	}
+
+	/* check task */
+	task = json_object_get_string_member (json_object, "Task");
+	if (g_strcmp0 (task, "upgrade") != 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Invalid task '%s', only 'upgrade' supported",
+			     task);
+		return FALSE;
+	}
+
+	/* find device */
+	device_id = json_object_get_string_member (json_object, "DeviceId");
+	device = fwupd_client_get_device_by_id (priv->client, device_id,
+						priv->cancellable, error);
+	if (device == NULL)
+		return FALSE;
+
+	/* find release for device */
+	checksum = json_object_get_string_member (json_object, "Checksum");
+	releases = fwupd_client_get_releases (priv->client, device_id,
+					      priv->cancellable, error);
+	for (guint i = 0; i < releases->len; i++) {
+		FwupdRelease *rel_tmp = g_ptr_array_index (releases, i);
+		if (fwupd_release_has_checksum (rel_tmp, checksum)) {
+			rel = g_object_ref (rel_tmp);
+			break;
+		}
+	}
+	if (rel == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Failed to find a release with a checksum of %s",
+			     checksum);
+		return FALSE;
+	}
+
+	/* FIXME: before I actually add this, is this really a good idea?! */
```

Comment:
The alternative being to rely upon a frontend actually doing the prompt and install?
My opinion here is that this turns into a question of how do you force updates onto a managed machine without potentially having data loss scenarios.

Right now there is no indication in fwupd whether it's "safe" to perform an update non-interactively.  This requires information on the topology of stuff behind it.

The example I'll give is if you have an update to a USB hub FW and have a USB disk attached to it.  The hub will restart into the new firmware, but that will also cause all the USB devices behind it to restart too.  If nothing was plugged in it's probably OK.  If a mouse is plugged in, it might annoy someone.  If a speaker is plugged in it might cause an app to crash if it was using that speaker.

Similarly if a graphics card was plugged into the Thunderbolt device, who knows what happens, graphics server crashes?

So my idea here is that all of the devices that are sent to update - check if they have `restart-required` set.  If they do, apply them immediately.  This should cover UEFI capsules, and disk updates.
If they don't have that set then mark them to be installed on the next boot or on logout.

Reply:
> set then mark them to be installed on the next boot

I think this is a textboox use of the "pending" next-boot functionality, right? I think if you're updating a machine remotely, for a user it would be odd for the mouse just to stop working, and then randomly start working 10 seconds later.

Reply:
Yes exactly.  But I'm saying to do the UEFI and ATA updates not as pending so that it's not an extra reboot just to reboot again.